got: 0.83 -> 0.86

afh opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

afh committed:
got: 0.83 -> 0.85

afh commented on pkgs/applications/version-management/got/default.nix:
```diff
@@ -17,12 +17,18 @@ stdenv.mkDerivation rec {
     substituteInPlace configure --replace 'xdarwin' 'xhomebrew'
   '';
 
+  patches = [ ./darwin.patch ];
```
Without this patch the buildPhase (linking) fails with missing symbol bsd_getopt

wahjava replied:
Probably should use an `if` here, instead of lib.optional. It's now broken on Darwin :/

afh commented on pkgs/applications/version-management/got/default.nix:
```diff
@@ -17,12 +17,18 @@ stdenv.mkDerivation rec {
 
+  patches = [ ./darwin.patch ];
+
+  preBuild = ''
+    buildFlagsArray+=(CFLAGS="-DHAVE_STRLCPY=1 -DHAVE_STRLCAT=1 -DHAVE_STRNSTR=1 -DHAVE_STRMODE=1")
+  '';
```
Without these buildFlags the buildPhase fails on macOS with conflicting types for strmode.

wahjava commented:
Result of `nixpkgs-review pr 220891` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>got</li>
  </ul>
</details>

wahjava committed:
got: add workarounds for Linux platform

> /nix/store/f8f55yg3p69hzywnb4nk3c4b2i0x0h0d-got-0.85/bin/got: error while loading shared libraries: libmd.so.0: cannot open shared object file: No such file or directory

fmt_scaled.c:226:17: warning: implicit declaration of function 'strlcpy'; did you mean 'strncpy'? [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration8;;]
  226 |                 strlcpy(result, "0B", FMT_SCALED_STRSIZE);
      |                 ^~~~~~~
      |                 strncpy

wahjava reviewed (changes requested):
It doesn't quite build on GNU/Linux as is, without https://github.com/wahjava/nixpkgs/commit/54104f1a044ef0419158daac6a24296c183ce2e9. Could you please `git-am` it in your branch, and push it ?

afh commented:
I've applied the patch as you requested, @wahjava.

What are your thoughts on gating the darwin patch in [libbsd](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/libbsd/default.nix#L25) with `lib.optional stdenv.isDarwin`?

wahjava commented:
Result of `nixpkgs-review pr 220891` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>got</li>
  </ul>
</details>

wahjava commented:
> I've applied the patch as you requested, @wahjava.
> 
> What are your thoughts on gating the darwin patch in [libbsd](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/libbsd/default.nix#L25) with `lib.optional stdenv.isDarwin`?

I don't have any strong opinions on that. AFAICT, if the functions in the question are already indicated as defined, then don't re-define them, which is fine.

In your commit you were marking those functions as being available (for GNU/Linux) when they weren't.

afh commented:
Thanks, @wahjava, I'll look into it.

afh committed:
got: Replace lib.optinals with lib.optionalString

where appropriate

afh committed:
got: Apply Darwin specific patch only on Darwin

afh committed:
got: 0.85 -> 0.86

wahjava reviewed (approved):
Looks good to me. Thank you!

afh commented:
Thanks for the review, @wahjava, much appreciated 🙏  Looking forward to having this merged.

wahjava commented:
Many thanks for working on it :bow:

wegank commented:
Superseded by #222170.
